main.go

```diff
@@ -106,38 +111,66 @@ func main() {
+	flagset.StringVar(&cfg.auth.Authentication.OIDC.ClientID, "oidc-clientID", "", "The client ID for the OpenID Connect client, must be set if oidc-issuer-url is set.")
+	flagset.StringVar(&cfg.auth.Authentication.OIDC.GroupsClaim, "oidc-groups-claim", "groups", "Identifier of groups in JWT claim, by default set to 'groups'")
+	flagset.StringVar(&cfg.auth.Authentication.OIDC.UsernameClaim, "oidc-username-claim", "email", "Identifier of the user in JWT claim, by default set to 'email'")
+	flagset.StringVar(&cfg.auth.Authentication.OIDC.GroupsPrefix, "oidc-groups-prefix", "", "If provided, all groups will be prefixed with this value to prevent conflicts with other authentication strategies.")
+	flagset.StringArrayVar(&cfg.auth.Authentication.OIDC.SupportedSigningAlgs, "oidc-sign-alg", []string{"RS256"}, "Supported signing algorithms, default RS256")
+	flagset.StringVar(&cfg.auth.Authentication.OIDC.CAFile, "oidc-ca-file", "", "If set, the OpenID server's certificate will be verified by one of the authorities in the oidc-ca-file, otherwise the host's root CA set will be used.")
+
+	//Kubeconfig flag
+	flagset.StringVar(&cfg.kubeconfigLocation, "kubeconfig", "", "Path to a kubeconfig file, specifying how to connect to the API server. If unset, in-cluster configuration will be used")
+	
 	flagset.Parse(os.Args[1:])
+	kcfg := initKubeConfig(cfg.kubeconfigLocation)
 
 	upstreamURL, err := url.Parse(cfg.upstream)
 	if err != nil {
 		glog.Fatalf("Failed to build parse upstream URL: %v", err)
 	}
 
-	kcfg, err := rest.InClusterConfig()
-	if err != nil {
-		glog.Fatalf("Failed to build Kubernetes rest-config: %v", err)
-	}
-
 	kubeClient, err := kubernetes.NewForConfig(kcfg)
 	if err != nil {
 		glog.Fatalf("Failed to instantiate Kubernetes client: %v", err)
 	}
 
-	if cfg.resourceAttributesFile != "" {
-		b, err := ioutil.ReadFile(cfg.resourceAttributesFile)
+	var authenticator authenticator.Request
+	// If OIDC configuration provided, use oidc authenticator
+	if cfg.auth.Authentication.OIDC.IssuerURL != "" {
+		authenticator, err = authn.NewOIDCAuthenticator(cfg.auth.Authentication.OIDC)
 		if err != nil {
-			glog.Fatalf("Failed to read resource-attribute file: %v", err)
+			glog.Fatalf("Failed to instantiate OIDC authenticator: %v", err)
 		}
 
-		err = yaml.Unmarshal(b, &cfg.auth.Authorization.ResourceAttributes)
-		if err != nil {
-			glog.Fatalf("Failed to parse resource-attribute file content: %v", err)
+	} else {
+		//Use Delegating authenticator
+		if kubeClient != nil && !reflect.ValueOf(kubeClient).IsNil() {
+			tokenClient := kubeClient.AuthenticationV1beta1().TokenReviews()
+			authenticator, err = authn.NewDelegatingAuthenticator(tokenClient, cfg.auth.Authentication)
+			if err != nil {
+				glog.Fatalf("Failed to instantiate delegating authenticator: %v", err)
+			}
 		}
 	}
 
-	auth, err := auth.BuildAuthHandler(kubeClient, &cfg.auth)
+	if kubeClient == nil && reflect.ValueOf(kubeClient).IsNil() {
```

Comment:
Is this check really necessary? `initKubeConfig` should err if there is no kubeClient available.